From 62b5126f2cf6a081557fa6f06554a8cd60de0c19 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tom=C3=A1s=20Gonz=C3=A1lez?=
 <tomasagustin.gonzalezorlando@arm.com>
Date: Tue, 14 Mar 2023 11:15:09 +0000
Subject: [PATCH] net: zperf: unify get_session among TCP and UDP

The modules zperf_udp_receiver and zperf_tcp_receiver uses two
different function to get a zperf session where to store the
statistics, there is a TODO comment in the zperf_session module
suggesting to unify that part.
So delete the get_tcp_session function and use get_session for
both TCP and UDP receiver module.
Delete sock field from struct session because it's not used
anymore.

Issue-Id: SCM-5823
Signed-off-by: Luca Fancellu <luca.fancellu@arm.com>
Upstream-Status: Backport [main:910de7f83d1e32a0f15fd1c4044045e8ada605d2]
---
 subsys/net/lib/zperf/zperf_session.c      | 36 -----------------------
 subsys/net/lib/zperf/zperf_session.h      |  3 --
 subsys/net/lib/zperf/zperf_tcp_receiver.c | 19 +++++++-----
 3 files changed, 12 insertions(+), 46 deletions(-)

diff --git a/subsys/net/lib/zperf/zperf_session.c b/subsys/net/lib/zperf/zperf_session.c
index bb195d7eb1b..de2dbf0230f 100644
--- a/subsys/net/lib/zperf/zperf_session.c
+++ b/subsys/net/lib/zperf/zperf_session.c
@@ -85,42 +85,6 @@ struct session *get_session(const struct sockaddr *addr,
 	return active;
 }
 
-/* TODO Unify session handling */
-struct session *get_tcp_session(int sock)
-{
-	struct session *free = NULL;
-	int i = 0;
-
-	if (sock < 0) {
-		NET_ERR("Error! Invalid socket.\n");
-		return NULL;
-	}
-
-	/* Check whether we already have an active session */
-	while (i < SESSION_MAX) {
-		struct session *ptr = &sessions[SESSION_TCP][i];
-
-		if (ptr->sock == sock) {
-			return ptr;
-		}
-
-		if (ptr->state == STATE_NULL ||
-		    ptr->state == STATE_COMPLETED) {
-			/* We found a free slot - just in case */
-			free = ptr;
-			break;
-		}
-
-		i++;
-	}
-
-	if (free) {
-		free->sock = sock;
-	}
-
-	return free;
-}
-
 void zperf_reset_session_stats(struct session *session)
 {
 	if (!session) {
diff --git a/subsys/net/lib/zperf/zperf_session.h b/subsys/net/lib/zperf/zperf_session.h
index 7e298c26eb5..c194d8a013d 100644
--- a/subsys/net/lib/zperf/zperf_session.h
+++ b/subsys/net/lib/zperf/zperf_session.h
@@ -36,9 +36,6 @@ struct session {
 	uint16_t port;
 	struct net_addr ip;
 
-	/* TCP session */
-	int sock;
-
 	enum state state;
 
 	/* Stat data */
diff --git a/subsys/net/lib/zperf/zperf_tcp_receiver.c b/subsys/net/lib/zperf/zperf_tcp_receiver.c
index 4dd53da8284..3cab08dc40a 100644
--- a/subsys/net/lib/zperf/zperf_tcp_receiver.c
+++ b/subsys/net/lib/zperf/zperf_tcp_receiver.c
@@ -52,14 +52,14 @@ static bool tcp_server_stop;
 static uint16_t tcp_server_port;
 static K_SEM_DEFINE(tcp_server_run, 0, 1);
 
-static void tcp_received(int sock, size_t datalen)
+static void tcp_received(const struct sockaddr *addr, size_t datalen)
 {
 	struct session *session;
 	int64_t time;
 
 	time = k_uptime_ticks();
 
-	session = get_tcp_session(sock);
+	session = get_session(addr, SESSION_TCP);
 	if (!session) {
 		NET_ERR("Cannot get a session!");
 		return;
@@ -233,6 +233,8 @@ static void tcp_server_session(void)
 	NET_INFO("Listening on port %d", tcp_server_port);
 
 	while (true) {
+		struct sockaddr addr_ipv4, addr_ipv6;
+
 		ret = zsock_poll(fds, ARRAY_SIZE(fds), POLL_TIMEOUT_MS);
 		if (ret < 0) {
 			NET_ERR("TCP receiver poll error (%d)", errno);
@@ -248,8 +250,8 @@ static void tcp_server_session(void)
 		}
 
 		for (int i = 0; i < ARRAY_SIZE(fds); i++) {
-			struct sockaddr addr;
-			socklen_t addrlen = sizeof(addr);
+			struct sockaddr *addr = &addr_ipv6;
+			socklen_t addrlen = sizeof(struct sockaddr);
 
 			if ((fds[i].revents & ZSOCK_POLLERR) ||
 			    (fds[i].revents & ZSOCK_POLLNVAL)) {
@@ -264,9 +266,10 @@ static void tcp_server_session(void)
 
 			switch (i) {
 			case SOCK_ID_IPV4_LISTEN:
+				addr = &addr_ipv4;
+				__fallthrough;
 			case SOCK_ID_IPV6_LISTEN:{
-				int sock = zsock_accept(fds[i].fd, &addr,
-							&addrlen);
+				int sock = zsock_accept(fds[i].fd, addr, &addrlen);
 
 				if (sock < 0) {
 					NET_ERR("TCP receiver IPv%d accept error",
@@ -292,6 +295,8 @@ static void tcp_server_session(void)
 			}
 
 			case SOCK_ID_IPV4_DATA:
+				addr = &addr_ipv4;
+				__fallthrough;
 			case SOCK_ID_IPV6_DATA:
 				ret = zsock_recv(fds[i].fd, buf, sizeof(buf), 0);
 				if (ret < 0) {
@@ -301,7 +306,7 @@ static void tcp_server_session(void)
 					goto error;
 				}
 
-				tcp_received(fds[i].fd, ret);
+				tcp_received(addr, ret);
 
 				if (ret == 0) {
 					zsock_close(fds[i].fd);
-- 
2.39.1

